### ISSUE TITLE
No notifications

### ISSUE THREAD

**Issue by ghost** [nextcloud/talk-ios#115]
Hello,
I don't have any notifications when a chat comes on IOS. Notification apps in Nextcloud is enabled and Notification is allowed for Nextcloud Talk in IOS.

**Comment by Ivansss** [comment]
Hi @tomaaa ,
At the moment, there are only notifications for calls, invitations and chat mentions.

**Comment by ghost** [comment]
Hi Ivansss,
thanks for your reply. I don't get any notifications for calls also.

**Comment by Ivansss** [comment]
Ok... 🤔 
I read your post here: https://help.nextcloud.com/t/no-notifications-nextcloud-talk-ios-chat/32642
Did you try to log-out and log-in again?
Also, you could try to log-out, then remove all the sessions of the Nextcloud Talk app from the web UI (Settings->Security) and reinstall the iOS app.

**Comment by Patate2** [comment]
I tried that.  and.... no, nothing, niet, nada.               
I'm crying.

**Comment by mario** [comment]
Do you have the notifications app installed on the server?

**Comment by Ivansss** [comment]
@Patate2 could you create a test account in your instance and send it to `ivan AT nextcloud DOT com`?
So I could debug what could be wrong in the subscription process?

**Comment by mario** [comment]
The notifications are supposed to work out of the box as we use a Nextcloud-company-hosted service that sends the push messages through Google/Apple. 

All data sent is encrypted, as you can check in the notifications app.

**Comment by jsalort** [comment]
@mario: OK, I understand. Thanks for the clarification. 

So, if I understand, you send a [POST request](https://github.com/nextcloud/talk-ios/blob/d03a6ef4b84bc3e020f7f4ba05bd0e8f245fcd79/VideoCalls/NCAPIController.m#L698) to [kNCPushServer](https://github.com/nextcloud/talk-ios/blob/d03a6ef4b84bc3e020f7f4ba05bd0e8f245fcd79/VideoCalls/NCSettingsController.m#L35) to subscribe the device to the Nextcloud-company-hosted notification service at push-notifications.nextcloud.com.
I guess the Notification app running on the instance already knows about this, and takes care of sending push notification to push-notifications.nextcloud.com (even if it is not clearly stated in the Notification app description that requests are going to be sent to a third-party server).

So, to track down where the notification gets lost, I suppose it should be possible to provide you with some log history:
- Is there a way to log into this nextcloud-hosted service, in order to verify that the registration was indeed successful ?
- Are there log files written by the Notification app on the instance, where it would be possible to check that push requests are successfully sent to push-notifications.nextcloud.com ?

**Comment by mario** [comment]
@jsalort the push server URL is coded inside the respective apps (e.g. iOS in this instance), not in the Notifications app. As for the logs: You could connect your phone to xcode, delete the account, and then re-add it. You should see if the registration process works then I think.

**Comment by jsalort** [comment]
> the push server URL is coded inside the respective apps (e.g. iOS in this instance), not in the Notifications app.

@mario OK, that means that I still don't fully grasp how this works.
I get notified in a chat. This happens on the instance. The instance can't directly tell Apple Notification Service because it does not have credentials and certificate. The nextcloud-company-hosted service holds a certificate and pushes the notification to Apple Notification Service. How does the nextcloud-company-hosted service know that I was notified in a chat ?

> You could connect your phone to xcode, delete the account, and then re-add it. You should see if the registration process works then I think.

OK. I am not sure what to look for, but it seems that everything is fine from this end:
```
VideoCalls	[com.nextcloud.Talk] Requesting token for remote notifications
VideoCalls	[com.nextcloud.Talk] Requested token for remote notifications [ didSucceed: 1 hasError: 0 hasCompletionHandler: 1 ]
VideoCalls	[com.nextcloud.Talk] Did receive remote notification token
VideoCalls	[com.nextcloud.Talk] Sending received device token to delegate UIApplication
VideoCalls	[com.nextcloud.Talk] Requested authorization [ didGrant: 1 hasError: 0 hasCompletionHandler: 1 ]

[...]

VideoCalls	4.13.0 - [Firebase/Messaging][I-FCM001000] FIRMessaging Remote Notifications proxy enabled, will swizzle remote notification receiver handlers. If you'd prefer to manually integrate Firebase Messaging, add "FirebaseAppDelegateProxyEnabled" to your Info.plist, and set it to NO. Follow the instructions at:
https://firebase.google.com/docs/cloud-messaging/ios/client#method_swizzling_in_firebase_messaging
to ensure proper integration.
VideoCalls	Push Notifications Key Pair generated: 
```

But is this expected that there is a direct connect to the notification API, and that it is "cancelled" (whatever that means):
```
VideoCalls	[3 <private> stream, pid: 15091, url: https://myserver-hidden.com/ocs/v2.php/apps/notifications/api/v2/push, tls] cancelled
	[3.1 67096779-F5D0-4277-B283-E5B1E04619FD <private>.59508<-><private>]
	Connected Path: satisfied (Path is satisfied), interface: en0, ipv4, ipv6, dns
	Duration: 0.552s, DNS @0.004s took 0.030s, TCP @0.051s took 0.016s, TLS took 0.018s
	bytes in/out: 1330/1674, packets in/out: 5/4, rtt: 0.026s, retransmitted packets: 0, out-of-order packets: 0
```

**Comment by mario** [comment]
For some reason (and I didn't code the iOS part, though I know most about push I guess) the registration process for push with your server fails (e.g. connection is cancelled). How did you login? Did you use the usual login flow?

The nextcloud-company-hosted-service has absolutely NO IDEA you were mentioned in a chat. All it knows is that your Nextcloud instance wants to send you (and others) some message. That message IS encrypted.

**Comment by jsalort** [comment]
> For some reason (and I didn't code the iOS part, though I know most about push I guess) the registration process for push with your server fails (e.g. connection is cancelled).

Ah interesting.
So I logged out and back in again, and this time I also tracked my Apache log.
And there it is:
```
 - [11/Jul/2018:23:34:21 +0200] "POST /index.php/login/flow HTTP/1.1" 303 1111 "-" "Columbus (Nextcloud Talk)"
 jsalort [11/Jul/2018:23:34:22 +0200] "POST /ocs/v2.php/apps/notifications/api/v2/push HTTP/1.1" 400 1303 "-" "Mozilla/5.0 (iOS) Nextcloud-Talk v2.1.0"
 jsalort [11/Jul/2018:23:34:22 +0200] "GET /ocs/v1.php/cloud/user?fomat=json HTTP/1.1" 200 1579 "-" "Mozilla/5.0 (iOS) Nextcloud-Talk v2.1.0"
 jsalort [11/Jul/2018:23:34:22 +0200] "GET /ocs/v2.php/apps/spreed/api/v1/room HTTP/1.1" 200 1552 "-" "Mozilla/5.0 (iOS) Nextcloud-Talk v2.1.0"
 jsalort [11/Jul/2018:23:34:22 +0200] "GET /ocs/v1.php/cloud/user?fomat=json HTTP/1.1" 200 1228 "-" "Mozilla/5.0 (iOS) Nextcloud-Talk v2.1.0"
```
It looks like login is fine, but the status code 400 on the POST request is presumably the cause of the malfunction... It would be interesting to know if others reporting the same issue have also this 400 status code. Could this be a configuration problem of my instance ? Or is it an issue in the way the iOS app builds this particular request ?

> How did you login? Did you use the usual login flow?

Yes. This is the usual login flow: the iOS app shows a webview asking me to grant access to the app, I click "Grant" and then enter my credentials. And I see the right POST request in the Apache log.

> The nextcloud-company-hosted-service has absolutely NO IDEA you were mentioned in a chat. All it knows is that your Nextcloud instance wants to send you (and others) some message. That message IS encrypted.

OK, I understand the encryption part. So, let me rephrase the question: how does the nextcloud-company-hosted service knows that my Nextcloud instance wants to send me (and others) some message ? There still needs to be some request sent from my instance to the nextcloud-company-hosted service, and the service has to accept it and relay...

**Comment by mario** [comment]
Error 400 means this:

Invalid device public key; device does not use a token to authenticate; the push token hash is invalid formatted; the proxy server URL is invalid;

@jsalort what version of Nc, what version of the notifications app?

As for the push:

- when you add an account, your device generates a device-specific keypair that is used to encrypt messages
- you register for push messages both at the Nextcloud server (your own) and at the nextcloud-company-hosted-push-magic-service
- Nextcloud server informs nextcloud-company-hosted-push-magic-service that you it wants to send a push message to one (or more) of the users on your server about a certain event

I hope that clarifies it!

**Comment by jsalort** [comment]
@mario thanks for your detailed answers.

> Error 400 means this:
>
> Invalid device public key; device does not use a token to authenticate; the push token hash is invalid formatted; the proxy server URL is invalid;
>
> @jsalort what version of Nc, what version of the notifications app?

**On the server**
- Ubuntu 16.04.4 LTS
- php 7.0.30
- mysql 5.7.22
- Nextcloud 13.0.4
- Notification app 2.1.2
- Talk app 3.2.3
- (no TURN server and no Signal server configured)

**On the iPhone**
- iOS 11.4 (15F79)
- Nextcloud Talk 2.1.0

> I hope that clarifies it!

It does. Thank you.

**Comment by jsalort** [comment]
@mario additional info while having a look at the logs on the nextcloud instance. I have several lines like this:
```
Error    notes    Moving note 3606 (VideoCalls [com.nextcloud.Talk] Requesting token for remote notifications) to the desired target has failed with a OCP\Files\InvalidPathException: Le nom de fichier contient un (des) caractère(s) non valide(s)
```
(the filename contains invalid characters)

**Comment by jsalort** [comment]
> Error 400 means this:
>
> Invalid device public key; device does not use a token to authenticate; the push token hash is invalid formatted; the proxy server URL is invalid;

Indeed, I've found that the details are explained here:
https://github.com/nextcloud/notifications/blob/master/docs/push-v2.md

So I have followed the readme and first checked if notification supports push on the server by sending a request to my instance on the endpoint `/ocs/v2.php/cloud/capabilities`
and I obtain this:
```
{'ocs': {'data': {'capabilities': {'bruteforce': {'delay': 0},
                                   'spreed': {'features': ['audio',
                                                           'video',
                                                           'chat-v2',
                                                           'guest-signaling',
                                                           'empty-group-room',
                                                           'guest-display-names',
                                                           'multi-room-users']},
                                   'theming': {'background': 'https://myserver.url/core/img/background.png?v=0',
                                               'background-default': True,
                                               'background-plain': False,
                                               'color': '#0082c9',
                                               'color-element': '#0082c9',
                                               'color-text': '#ffffff',
                                               'logo': 'https://myserver.url/core/img/logo.svg?v=0',
                                               'name': 'Nextcloud',
                                               'slogan': 'a safe home for all '
                                                         'your data',
                                               'url': 'https://nextcloud.com'}},
                  'version': {'edition': '',
                              'major': 13,
                              'micro': 4,
                              'minor': 0,
                              'string': '13.0.4'}},
         'meta': {'message': 'OK', 'status': 'ok', 'statuscode': 200}}}
```
So, there is no `"notification" : {"push": [... "devices"]}` in the capabilities.
Is there something that has to be activated on the nextcloud instance ?

Back on topic, I would like to make a suggestion for the Nextcloud Talk iOS app. I don't know if I should open another issue (enhancement request). I could have tried to implement it and make a PR but I am not familiar enough with iOS programming and I haven't even been able to build talk-ios app myself (Xcodes complains that my development team can't create push notifications and that it cannot create a iOS App Development provisioning profile for "com.nextcloud.Talk".):
- the talk-ios app should first check the capabilities of the Nextcloud server, and report the missing support for push on the server;
- on subscription failure, talk-ios app shouldn't silently skip subscription, but report the error to the user and print the `message` body of the answer (in case of code `400`), i.e. `INVALID_PUSHTOKEN_HASH`, `INVALID_SESSION_TOKEN`, `INVALID_DEVICE_KEY` or `INVALID_PROXY_SERVER`. That would make it easier to figure out what is going wrong.

**Comment by mario** [comment]
Yes, notifications app? :D

**Comment by jsalort** [comment]
> Yes, notifications app? :D

The Notifications app is installed and activated. Besides, I get notifications in the web UI.
The symptoms are similar to what @tomaa and @Patate2 describe.

This is apparently a Notifications app issue, rather than a Nextcloud Talk iOS app issue. So I have filed an issue there: https://github.com/nextcloud/notifications/issues/143

However, the feature enhancement request holds for the Nextcloud Talk iOS app to get some informative message when notification subscription does not work. Shall I open a separate issue with this enhancement request ?

**Comment by jsalort** [comment]
> Yes, notifications app? :D

Unfortunately, this does not seem to be a problem with the Notifications app. One has to be authenticated to get the list of capabilities at the `/ocs/v2.php/cloud/capabilities` endpoint.
When authenticated, I get "notifications" in the list of capabilities.

Therefore, it is really the Talk iOS app which somehow sends a bad request...

**Comment by jsalort** [comment]
I have updated to Nextcloud 14, Nextcloud Talk 4.0.0 and NextCloud Talk iOS 3.0.1. It stills produces HTTP code 400 and an error in iOS console:
```
par défaut	22:20:15.800350 +0200	VideoCalls	Error while subscribing to NC server.
par défaut	22:20:15.800408 +0200	VideoCalls	[3 <private> stream, pid: 7807, url: https://myserver-hidden/ocs/v2.php/apps/notifications/api/v2/push, tls] cancelled
	[3.1 09F8D5CB-259A-45D4-B7B7-B1DC606BB4D9 <private>.63554<-><private>]
	Connected Path: satisfied (Path is satisfied), interface: en0, ipv4, dns
	Duration: 0.484s, DNS @0.006s took 0.051s, TCP @0.064s took 0.021s, TLS took 0.021s
	bytes in/out: 1456/2520, packets in/out: 4/5, rtt: 0.019s, retransmitted packets: 1, out-of-order packets: 0
```
The first message comes from this NSLog line:
https://github.com/nextcloud/talk-ios/blob/c57981217c19dd6377ef0a6009537e26b53ee772/VideoCalls/AuthenticationViewController.m#L150

I would like to add an additional informative console log message, if that is possible. So I have submitted a PR: https://github.com/nextcloud/talk-ios/pull/152

**Comment by Ivansss** [comment]
Hi @jsalort :)
Could you send me a test account for you server to `ivan AT nextcloud.com`?

**Comment by Ivansss** [comment]
The error while subscribing to the NC server has been fixed in the server.
https://github.com/nextcloud/server/pull/11302
Thanks @jsalort for your cooperation :)
